compiler/src/dotty/tools/dotc/typer/Deriving.scala

```diff
@@ -83,63 +83,170 @@ trait Deriving { this: Typer =>
      *  that have the same name but different prefixes through selective aliasing.
      */
     private def processDerivedInstance(derived: untpd.Tree): Unit = {
-      val originalType = typedAheadType(derived, AnyTypeConstructorProto).tpe
-      val underlyingType = underlyingClassRef(originalType)
-      val derivedType = checkClassType(underlyingType, derived.sourcePos, traitReq = false, stablePrefixReq = true)
-      val typeClass = derivedType.classSymbol
-      val nparams = typeClass.typeParams.length
-
-      lazy val clsTpe = cls.typeRef.EtaExpand(cls.typeParams)
-      if (nparams == 1 && clsTpe.hasSameKindAs(typeClass.typeParams.head.info)) {
-        // A "natural" type class instance ... the kind of the data type
-        // matches the kind of the unique type class type parameter
-
-        val resultType = derivedType.appliedTo(clsTpe)
-        val instanceInfo = ExprType(resultType)
-        addDerivedInstance(originalType.typeSymbol.name, instanceInfo, derived.sourcePos)
-      } else {
-        // A matrix of all parameter combinations of current class parameters
-        // and derived typeclass parameters.
-        // Rows: parameters of current class
-        // Columns: parameters of typeclass
-
-        // Running example: typeclass: class TC[X, Y, Z], deriving class: class A[T, U]
+      val originalTypeClassType = typedAheadType(derived, AnyTypeConstructorProto).tpe
+      val typeClassType = checkClassType(underlyingClassRef(originalTypeClassType), derived.sourcePos, traitReq = false, stablePrefixReq = true)
+      val typeClass = typeClassType.classSymbol
+
+      def sameParamKinds(xs: List[ParamInfo], ys: List[ParamInfo]): Boolean =
+        xs.corresponds(ys)((x, y) => x.paramInfo.hasSameKindAs(y.paramInfo))
+
+      def cannotBeUnified =
+        ctx.error(i"${cls.name} cannot be unified with the type argument of ${typeClass.name}", derived.sourcePos)
+
+      def addInstance(derivedParams: List[TypeSymbol], evidenceParamInfos: List[List[Type]], instanceTypes: List[Type]): Unit = {
+        val resultType = typeClassType.appliedTo(instanceTypes)
+        val methodOrExpr =
+          if (evidenceParamInfos.isEmpty) ExprType(resultType)
+          else ImplicitMethodType(evidenceParamInfos.map(typeClassType.appliedTo), resultType)
+        val derivedInfo = if (derivedParams.isEmpty) methodOrExpr else PolyType.fromParams(derivedParams, methodOrExpr)
+        addDerivedInstance(originalTypeClassType.typeSymbol.name, derivedInfo, derived.sourcePos)
+      }
+
+      val typeClassParams = typeClass.typeParams
+      val typeClassArity = typeClassParams.length
+      if (typeClassArity == 1) {
+        // Primary case: single parameter type classes
```

Comment:
It's a rather long code sequence under this condition. Can we move it into a separate method?